File: src/cmdbox/repositories/tag_repository.py
from typing import Sequence

from peewee import IntegrityError

from .base_repository import BaseRepository
from .errors import NameConflictError, UnknownNameError, ValidationError
from .validators import TagValidator
from cmdbox.models import Tag


class TagRepository(BaseRepository[Tag]):
    model = Tag

    def __init__(self, validator: TagValidator | None = None):
        self.validator = validator or TagValidator()

    def create(self, name: str, description: str | None = None) -> Tag:
        name = name.strip() if name else None
        self.validator.validate_create(name=name, description=description)
        try:
            return Tag.create(name=name, description=description)
        except IntegrityError as exc:
            if name is not None and self._is_unique_name_violation(exc):
                raise NameConflictError(name=name) from exc
            raise

    def get_by_name(self, name: str) -> Tag | None:
        name = name.lower()
        tag = Tag.get_or_none(Tag.name == name)
        if tag is None:
            raise UnknownNameError(name=name)
        return tag

    def update(self, tag_name: str, **fields) -> Tag | None:
        tag = self.get_by_name(tag_name)
        if not fields:
            raise ValueError("No fields provided for update.")

        if "name" in fields and fields.get("name") is not None:
            fields["name"] = fields.get("name").strip()

        self.validator.validate_update(
            name=fields.get("name", tag.name),
            description=fields.get("description", tag.description),
        )
        try:
            for key, value in fields.items():
                if not hasattr(tag, key):
                    raise ValidationError(f"Invalid field: {key}")
                if value is not None:
                    setattr(tag, key, value)
            tag.save()
            return tag
        except IntegrityError as exc:
            name = fields.get("name", "")
            if name is not None and self._is_unique_name_violation(exc):
                raise NameConflictError(name=name) from exc
            raise

    def list_all(
        self, order_by: str | Sequence[str] = "name", limit: int = 25
    ) -> list[Tag]:
        ordering = self._resolve_ordering(order_by)
        return list(Tag.select().order_by(*ordering).limit(limit))

    def search(
        self, query: str, fields: str | Sequence[str] | None = ("name", "description")
    ) -> list[Tag]:
        return self._search(query, "name", fields)

    def delete(self, name: str) -> bool:
        tag = self.get_by_name(name)
        if not tag:
            return False
        tag.delete_instance()
        return True

File: internal/repository/tag_repository.go
package repository

import (
	"errors"
	"fmt"
	"strings"

	"github.com/PhantomLambSoft/CmdBox/internal/models"
	"github.com/PhantomLambSoft/CmdBox/internal/repository/validate"
	"gorm.io/gorm"
)

var tagOrderableColumns = map[string]bool{
	"id":          true,
	"name":        true,
	"description": true,
	"created_at":  true,
	"updated_at":  true,
}

var defaultTagSearchFields = []string{"name", "description"}

type TagCreateConfig struct {
	Name        string
	Description *string
}

type TagUpdateConfig struct {
	Name        *string
	Description *string
}

type TagRepository interface {
	Create(input TagCreateConfig) (*models.Tag, error)
	GetByName(name string) (*models.Tag, error)
	GetByID(ID uint) (*models.Tag, error)
	Update(tag *models.Tag, input TagUpdateConfig) (*models.Tag, error)
	Delete(command *models.Tag) error

	ListAll(orderBy string, limit int) ([]models.Tag, error)
	Search(query string, fields []string, limit int) ([]models.Tag, error)
}

type tagRepository struct {
	db        *gorm.DB
	validator *validate.TagValidator
}

func NewTagRepository(db *gorm.DB, validator *validate.TagValidator) TagRepository {
	if validator == nil {
		validator = validate.NewTagValidator(nil)
	}
	return &tagRepository{db: db, validator: validator}
}

func (r *tagRepository) Create(input TagCreateConfig) (*models.Tag, error) {
	name := strings.TrimSpace(input.Name)
	if err := r.validator.ValidateCreate(name); err != nil {
		return nil, err
	}

	tag := &models.Tag{
		Name:        name,
		Description: input.Description,
	}

	if err := r.db.Create(tag).Error; err != nil {
		if isUniqueConstraintViolation(err, "tags", "name") {
			return nil, fmt.Errorf("%w: %q", ErrTagNameConflict, name)
		}
		return nil, fmt.Errorf("creating tag: %q: %w", name, err)
	}
	return tag, nil
}

func (r *tagRepository) GetByName(name string) (*models.Tag, error) {
	var tag models.Tag
	err := r.db.Where("name = ?", name).First(&tag).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, fmt.Errorf("%w: %q", ErrUnknownTagName, name)
	}
	if err != nil {
		return nil, fmt.Errorf("getting tag by name: %q: %w", name, err)
	}
	return &tag, nil
}

func (r *tagRepository) GetByID(ID uint) (*models.Tag, error) {
	var tag models.Tag
	err := r.db.Where("id = ?", ID).First(&tag).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, fmt.Errorf("%w: %d", ErrUnknownTag, ID)
	}
	if err != nil {
		return nil, fmt.Errorf("getting tag by id: %d: %w", ID, err)
	}
	return &tag, nil
}

func (r *tagRepository) Update(tag *models.Tag, input TagUpdateConfig) (*models.Tag, error) {
	if tag == nil {
		return nil, ErrNoUpdateTarget
	}

	mergedName := tag.Name
	if input.Name != nil {
		mergedName = strings.TrimSpace(*input.Name)
	}

	if err := r.validator.ValidateUpdate(&mergedName); err != nil {
		return nil, err
	}

	if input.Name != nil {
		tag.Name = mergedName
	}
	if input.Description != nil {
		tag.Description = input.Description
	}

	if err := r.db.Save(tag).Error; err != nil {
		if isUniqueConstraintViolation(err, "tags", "name") {
			return nil, fmt.Errorf("%w: %q", ErrTagNameConflict, mergedName)
		}
		return nil, fmt.Errorf("updating tag: %d: %w", tag.ID, err)
	}
	return tag, nil
}

func (r *tagRepository) Delete(tag *models.Tag) error {
	if tag == nil {
		return nil
	}
	if err := r.db.Delete(tag).Error; err != nil {
		return fmt.Errorf("deleting tag: %d: %w", tag.ID, err)
	}
	return nil
}

func (r *tagRepository) ListAll(orderBy string, limit int) ([]models.Tag, error) {
	if orderBy == "" {
		orderBy = "name"
	}
	if limit <= 0 {
		limit = 25
	}
	orderClauses, err := resolveOrdering(orderBy, tagOrderableColumns)
	if err != nil {
		return nil, err
	}

	var tags []models.Tag
	q := r.db.Model(&models.Tag{})
	for _, clause := range orderClauses {
		q = q.Order(clause)
	}
	if err := q.Limit(limit).Find(&tags).Error; err != nil {
		return nil, fmt.Errorf("listing tags: %w", err)
	}
	return tags, nil
}

func (r *tagRepository) Search(query string, fields []string, limit int) ([]models.Tag, error) {
	if len(fields) == 0 {
		fields = defaultTagSearchFields
	}
	if limit <= 0 {
		limit = 25
	}

	var tags []models.Tag
	searchInput := searchWithRelevanceInput{
		DB:                   r.db,
		Table:                "tags",
		Query:                query,
		Fields:               fields,
		AllowedColumns:       tagOrderableColumns,
		SecondaryOrderColumn: "name",
		Limit:                limit,
		ExtraWhere:           "",
		ExtraArgs:            nil,
		Dest:                 &tags,
	}
	err := searchWithRelevance(searchInput)
	if err != nil {
		return nil, fmt.Errorf("searching tags: %w", err)
	}
	return tags, nil
}
